File: dlang/plugin-api/src/main/java/io/github/intellij/dlanguage/stubs/types/VersionSpecificationElementType.java
package io.github.intellij.dlanguage.stubs.types;

import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import com.intellij.util.io.StringRef;
import io.github.intellij.dlanguage.psi.named.DlangVersionSpecification;
import io.github.intellij.dlanguage.psi.impl.named.DlangVersionSpecificationImpl;
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes;
import io.github.intellij.dlanguage.stubs.VersionSpecificationStub;
import java.io.IOException;
import org.jetbrains.annotations.NotNull;

/**
 * Created by francis on 1/5/2018.
 */
public class VersionSpecificationElementType extends
    DNamedStubElementType<VersionSpecificationStub, DlangVersionSpecification> {

    public VersionSpecificationElementType(final String debugName) {
        super(debugName);
    }

    @Override
    public DlangVersionSpecification createPsi(@NotNull final VersionSpecificationStub stub) {
        return new DlangVersionSpecificationImpl(stub, this);
    }

    @NotNull
    @Override
    public VersionSpecificationStub createStub(@NotNull final DlangVersionSpecification psi,
        final StubElement parentStub) {
        return new VersionSpecificationStub(parentStub, this, psi.getName(), psi.getAttributes());
    }

    @NotNull
    @Override
    public VersionSpecificationStub deserialize(@NotNull final StubInputStream dataStream,
        final StubElement parentStub) throws IOException {
        final StringRef name = dataStream.readName();
        final DAttributes attributes = DAttributes.Companion.read(dataStream);
        return new VersionSpecificationStub(parentStub, this, name.getString(), attributes);

    }

    @Override
    public void serialize(@NotNull final VersionSpecificationStub stub,
        @NotNull final StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
        stub.getAttributes().write(dataStream);
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/VersionSpecificationElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import io.github.intellij.dlanguage.psi.impl.named.DlangVersionSpecificationImpl
import io.github.intellij.dlanguage.psi.named.DLanguageVersionSpecification
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes.Companion.read
import io.github.intellij.dlanguage.stubs.DLanguageVersionSpecificationStub
import java.io.IOException

/**
 * Created by francis on 1/5/2018.
 */
class VersionSpecificationElementType(debugName: String) :
    DNamedStubElementType<DLanguageVersionSpecificationStub, DLanguageVersionSpecification>(debugName) {
    override fun createPsi(stub: DLanguageVersionSpecificationStub): DLanguageVersionSpecification {
        return DlangVersionSpecificationImpl(stub, this)
    }

    override fun createStub(
        psi: DLanguageVersionSpecification,
        parentStub: StubElement<*>?
    ): DLanguageVersionSpecificationStub {
        return DLanguageVersionSpecificationStub(parentStub, this, psi.name, psi.attributes)
    }

    @Throws(IOException::class)
    override fun deserialize(
        dataStream: StubInputStream,
        parentStub: StubElement<*>?
    ): DLanguageVersionSpecificationStub {
        val name = dataStream.readName()
        val attributes = read(dataStream)
        return DLanguageVersionSpecificationStub(parentStub, this, name?.getString(), attributes)
    }

    @Throws(IOException::class)
    override fun serialize(
        stub: DLanguageVersionSpecificationStub,
        dataStream: StubOutputStream
    ) {
        dataStream.writeName(stub.name)
        stub.attributes.write(dataStream)
    }
}
